### ISSUE TITLE
Getting error: "adapter: poppler Error: Broken pipe"

### ISSUE THREAD

**Issue by bbardsley** [phiresky/ripgrep-all#113]
After the command rga --files-with-matches "search-term", I get for almost all the pdf files I have in my directory the error:

```
name-of-file.pdf:
adapter: poppler
Error: Broken pipe (os error 32)
```

For some files it also says:

```
adapter: poppler
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }', src/adapters/spawning.rs:93:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on
an `Err` value: Any { .. }', src/adapters/spawning.rs:98:6

```

I have no idea what causes this error.

**Comment by phiresky** [comment]
please send an example pdf file and show the output of `pdftotext -v`

**Comment by bbardsley** [comment]
The output of `pdftotext -v` is
`pdftotext version 21.08.0
Copyright 2005-2021 The Poppler Developers - http://poppler.freedesktop.org
Copyright 1996-2011 Glyph & Cog, LLC`

The error happens for example for
https://www.maths.ed.ac.uk/~v1ranick/papers/witten.pdf

**Comment by phiresky** [comment]
I can't reproduce the issue, for me `rga foo` in a directory with only the above pdf works fine. does `rga-preproc witten.pdf` work for you?

**Comment by bbardsley** [comment]
`rga foo` works fine for me. `rga --files-with-matches foo` also works fine, but somehow `rga --files-with-matches thing` gives an error. `rga-preproc witten.pdf` gives a normal output.

**Comment by TwistingTwists** [comment]
I am getting same error.

1.` rga locust`   -- works
2. ` rga locust --files-with-matches`  -- gives error 
```
adapter: poppler
Error: Broken pipe (os error 32)
```

**Comment by akobel** [comment]
`adapter: poppler` is a red herring, I guess: the problem also occurs for cached data, where pdftotext is not called at all.
The problem occurs quite randomly (0.9.6 on an up-to-date Arch installation), it's not pinned to any specific document. Looks like a race condition to me in the stream handling of `pipe_output` (not the cache lookup, I guess; I first suspected an issue in the zstd decoder handling, but I could reproduce also with clean cache).

However, the problem seems gone in master, where `pipe_output` has been reworked fundamentally, as far as I can tell, although the [comment regarding threading](https://github.com/phiresky/ripgrep-all/blob/2d63efd3156a2eca633b1b43d67931fe1cb0df6e/src/adapters/spawning.rs#L83) doesn't sound *too* stable yet...

**Comment by ygm-nob** [comment]
I use like this without using --files-with-matches:. I'm using fish shell on Mac. 

#!/opt/homebrew//bin/fish
/usr/bin/open (rga $argv[1] |  awk 'BEGIN{FS=":"}{print $1}' | sort -u | fzf --sort )

**Comment by mathieupost** [comment]
I'm having the exact same issue as @Ornanovitch in https://github.com/phiresky/ripgrep-all/issues/113#issuecomment-964347586 

Getting this error for every pdf at the first try (or when running with `--rga-no-cache`)
```
adapter: poppler
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }', src/adapters/spawning.rs:93:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', src/adapters/spawning.rs:98:6
```

Running with `RUST_BACKTRACE=1`:
```
adapter: poppler
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }

   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: __pthread_start
', src/adapters/spawning.rs:93:53
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: crossbeam_utils::thread::ScopedThreadBuilder::spawn::{{closure}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', src/adapters/spawning.rs:98:6
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: ripgrep_all::adapters::spawning::pipe_output
   4: ripgrep_all::adapters::spawning::<impl ripgrep_all::adapters::FileAdapter for T>::adapt
   5: ripgrep_all::preproc::rga_preproc
   6: rga_preproc::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

After running without `--files-with-matches` every run random pdf's get this error:
```
adapter: poppler
Error: Broken pipe (os error 32)
```

```
❯ rga --version
ripgrep-all 0.9.6

❯ rg --version
ripgrep 13.0.0
-SIMD -AVX (compiled)
+SIMD +AVX (runtime)
```

Both installed from `nixpkgs-22.05` on macOS

**Comment by p-sin** [comment]
> I am getting same error.
> 
> 1.` rga locust` -- works 2. ` rga locust --files-with-matches` -- gives error
> 
> ```
> adapter: poppler
> Error: Broken pipe (os error 32)
> ```

I had this same issue with the version provided on the ArchLinux repository (0.9.6). Installing from master branch solved the issue.

However the master branch has the peculiarity of not showing page-numbers with the results, for some reason. So I (admittedly) duct-taped a "solution" (or to be more honest: a hack) where my rga-fzf command uses a locally stored version of the rga from the master branch to list the files with `--files-with-matches`, and the latest release to do the actual search... Not proud of it, but it works currently.

**Comment by TwistingTwists** [comment]
can you share your rga-fzf function ? 

and how to get those two different versions of rga you mentioned ?

Thanks !

> > I am getting same error.
> > 1.` rga locust` -- works 2. ` rga locust --files-with-matches` -- gives error
> > ```
> > adapter: poppler
> > Error: Broken pipe (os error 32)
> > ```
> 
> I had this same issue with the version provided on the ArchLinux repository (0.9.6). Installing from master branch solved the issue.
> 
> However the master branch has the peculiarity of not showing page-numbers with the results, for some reason. So I (admittedly) duct-taped a "solution" (or to be more honest: a hack) where my rga-fzf command uses a locally stored version of the rga from the master branch to list the files with `--files-with-matches`, and the latest release to do the actual search... Not proud of it, but it works currently.

**Comment by jonaustin** [comment]
Not the original poster, but the way I did it is super simple:

```
git clone https://github.com/phiresky/ripgrep-all.git
cd ripgrep-all
cargo install --locked --path . --force # install master
pacman -S ripgrep-all # 0.9.6
```

the rga-fzf shell function is exactly the same as the one in this readme's repo; except the rga command uses master
```
rga-fzf () {
	RG_PREFIX="~/.cargo/bin/rga --files-with-matches"
	local file
	file="$(FZF_DEFAULT_COMMAND="$RG_PREFIX '$1'" \
    fzf --sort --preview="[[ ! -z {} ]] && rga --pretty --context 5 {q} {}" \
    --phony -q "$1" \
    --bind "change:reload:$RG_PREFIX {q}" \
    --preview-window="70%:wrap"
    )"  && echo "opening $file" && xdg-open "$file"
}```

I poked at the code at bit to see why it no longer works; looks like there was a major refactoring at some point and the poppler adapter has now moved into adapters/custom.rs for some reason; and this functionality just hasn't been added back yet:
```
            // postprocessors: [{name: "add_page_numbers_by_pagebreaks"}]
```
https://github.com/phiresky/ripgrep-all/blob/master/src/adapters/custom.rs

old code:
https://github.com/phiresky/ripgrep-all/blob/291033bf9f125af778e9008b8d9b346825dcc327/src/adapters/poppler.rs#L58

**Comment by temberature** [comment]
> Not the original poster, but the way I did it is super simple:
> 
> ```
> git clone https://github.com/phiresky/ripgrep-all.git
> cd ripgrep-all
> cargo install --locked --path . --force # install master
> pacman -S ripgrep-all # 0.9.6
> ```
> 
> the rga-_fzf_ shell function is exactly the same as the one in this readme's repo; except the first rga command uses master
> 
> ```
> rga-_fzf_ () {
>     RG_PREFIX="~/.cargo/bin/rga --files-with-matches"
>     local file
>     file="$(_FZF__DEFAULT_COMMAND="$RG_PREFIX '$1'" \
>     _fzf_ --sort --preview="[[ ! -z {} ]] && rga --pretty --context 5 {q} {}" \
>     --phony -q "$1" \
>     --bind "change:reload:$RG_PREFIX {q}" \
>     --preview-window="70%:wrap"
>     )"  && echo "opening $file" && xdg-open "$file"
> }
> ```
> 
> I poked at the code at bit to see why it no longer works; looks like there was a major refactoring at some point and the poppler adapter has now moved into adapters/custom.rs for some reason; and this functionality just hasn't been added back yet:
> 
> ```
>             // postprocessors: [{name: "add_page_numbers_by_pagebreaks"}]
> ```
> 
> https://github.com/phiresky/ripgrep-all/blob/master/src/adapters/custom.rs
> 
> old code:
> 
> https://github.com/phiresky/ripgrep-all/blob/291033bf9f125af778e9008b8d9b346825dcc327/src/adapters/poppler.rs#L58

I build from tag v0.9.6 like this and error disappear without behavior change.

**Comment by benthamite** [comment]
Have others who reported the issue found a solution to it? I just cloned the repo and built,

``` shell
git clone https://github.com/phiresky/ripgrep-all.git
cd ripgrep-all
cargo install --locked --path . --force # install master
```

but the issue persists. E.g. `rga --files-with-matches foo` returns

``` shell
adapter: poppler
Error: Broken pipe (os error 32)
```

System Version:	macOS 12.6 (21G115)

**Comment by mpr1255** [comment]
FWIW for anyone else who comes across this. I'm on OSX 12.6 and it wasn't working after installing with brew (including uninstalling & reinstalling everything). But jonaustin's 30 Jun post did the job: clone the repo, install with cargo. It even added it to the path and everything so 'rga' at the terminal just works. rga-fzf too.

**Comment by Bellavene** [comment]
brew install rga --HEAD

**Comment by juffis** [comment]
I just ran into this, too. Any progress on this?

**Comment by Bellavene** [comment]
What progress do you need, if you always can install the recent (--HEAD) version?

**Comment by jonaustin** [comment]
> What progress do you need, if you always can install the recent (--HEAD) version?

Using the function in this comment is still better because the current version still doesn't return page numbers with the search: https://github.com/phiresky/ripgrep-all/issues/113#issuecomment-1171250577

**Comment by Bellavene** [comment]
Thank you for letting know.

**Comment by juffis** [comment]
I'm trying to get this to work in termux as I have all my pdfs on my phone. 
 
> Using the function in this comment is still better because the current version still doesn't return page numbers with the search: [#113 (comment)](https://github.com/phiresky/ripgrep-all/issues/113#issuecomment-1171250577)

Do you mean the function with two different versions as posted by you?

**Comment by jonaustin** [comment]
Yeah (I mean it's better if you want the page numbers for each search result anyway).

No idea about termux. Neat though; post back if/how you get it working.

**Comment by phiresky** [comment]
Fixed in 0.10
